File: 13_Roman_to_Integer/13.py
def romanToInt(s):
    single_val = {'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000}
    special_val = {'IV': 4, 'IX': 9, 'XL': 40, 'XC': 90, 'CD': 400, 'CM': 900}
    idx = 0
    val = 0
    while True:
        # print(f'idx = {idx}')
        if idx == len(s) - 1:
            # last char of s, no need to compare special case
            val += single_val[s[idx]]
            # break
            idx += 1
        else:
            # not the last char, consider special case
            temp = s[idx] + s[idx + 1]
            if temp in special_val.keys():
                val += special_val[temp]
                idx += 2
            else:
                val += single_val[s[idx]]
                idx += 1
        
        # print(f'val = {val}')
        if idx == len(s):
            break
    return val



if __name__ == '__main__':    
    print(romanToInt('LVIII'))

File: 13_Roman_to_Integer/13.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


int romanToInt(char * s)
{
    int i;
    /*
    for(i = 0; i < strlen(s); i++) {
        printf("%c ", s[i]);
    }
    */
    int idx = 0;
    int val = 0;
    while(1) {
        if(idx == strlen(s) - 1) {
            switch(s[idx]) {
            case 'I': val += 1; break;
            case 'V': val += 5; break;
            case 'X': val += 10; break;
            case 'L': val += 50; break;
            case 'C': val += 100; break;
            case 'D': val += 500; break;
            case 'M': val += 1000; break;
            default: val = val; break;
            }
            idx ++;
        }

	    else {
            char now = s[idx];
            char next = s[idx + 1];
            if(now == 'I') {
                if(next == 'V') {
                    val += 4;
                    idx += 2;
                }
                else if(next == 'X') {
                    val += 9;
                    idx += 2;
                }
                else {
                    val += 1;
                    idx += 1;
                }
            }


            else if(now == 'X') {
                if(next == 'L') {
                    val += 40;
                    idx += 2;
                }
                else if(next == 'C') {
                    val += 90;
                    idx += 2;
                }
                else {
                    val += 10;
                    idx += 1;
                }
            }


            else if(now == 'C') {
                if(next == 'D') {
                    val += 400;
                    idx += 2;
                }
                else if(next == 'M') {
                    val += 900;
                    idx += 2;
                }
                else {
                    val += 100;
                    idx += 1;
                }

            }
        

            else {
                switch(s[idx]) {
                    case 'I': val += 1; break;
                    case 'V': val += 5; break;
                    case 'X': val += 10; break;
                    case 'L': val += 50; break;
                    case 'C': val += 100; break;
                    case 'D': val += 500; break;
                    case 'M': val += 1000; break;
                    default: val = val; break;
                }
                idx ++;
            }
	    }
		


	    if (idx == strlen(s)) break;
	}
	return val;
	
}





int main()
{

    printf("%d\n", romanToInt("MCMXCIV"));
    // romanToInt("LVIII");
    return 0;

}

